x-pack/libbeat/management/blacklist.go

```diff
@@ -0,0 +1,188 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package management
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/x-pack/libbeat/management/api"
+)
+
+// ConfigBlacklist takes a ConfigBlocks object and filter it based on the given
+// blacklist settings
+type ConfigBlacklist struct {
+	patterns map[string]*regexp.Regexp
+}
+
+// ConfigBlacklistSettings holds a list of fields and regular expressions to blacklist
+type ConfigBlacklistSettings map[string]string
+
+// Unpack unpacks nested fields set with dot notation like foo.bar into the proper nesting
+// in a nested map/slice structure.
+func (f ConfigBlacklistSettings) Unpack(to interface{}) error {
+	m, ok := to.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("wrong type, expect map")
+	}
+
+	var expand func(key string, value interface{})
+
+	expand = func(key string, value interface{}) {
+		switch v := value.(type) {
+		case map[string]interface{}:
+			for k, val := range v {
+				expand(fmt.Sprintf("%v.%v", key, k), val)
+			}
+		case []interface{}:
+			for i := range v {
+				expand(fmt.Sprintf("%v.%v", key, i), v[i])
+			}
+		default:
+			m[key] = fmt.Sprintf("%s", value)
+		}
+	}
+
+	for k, val := range m {
+		expand(k, val)
+	}
+	return nil
+}
```

Comment:
What is this type + method supposed to do? The method unpacks the given map into the given map, which is only allocated temporarily? Plus, Unpack as supported by go-ucfg requires a pointer receiver.

The current implementation of the Unpack looks broken to me. Do we have a system test applying a custom blacklisted config?

Why do we need an extra ConfigBlacklistSettings and ConfigBlacklist? One is the settings and the other one is the filter/checker?

How about type names (using full names): `management.BlacklistConfig` and `management.BlacklistCheck`. As we might introduce a whitelist in the future I see a Checker interface and have a package named: `cfgcheck`.

Regarding the Unpack method:

```
type Config struct {  // full name is: blacklist.Config
  patterns map[string]match.Matcher
}

// Note: with this signature go-ucfg will check the config object can be converted to map[string]interface{}
func (s *Config) Unpack(from map[string]interface{}) (err error) {
  s.patterns, err = expandMatchers(from)
  return err
}

func expandMatchers(from map[string]interface{}) (map[string]*matcher.Match, error) {
  settings := flattenConfigs(from)
  patterns := make(map[string]*matcher.Match, len(settings))
  for k, v := range settings {
    m, err := match.Compile(fmt.Sprintf("%s", v))
    if err != nil {
      return nil, errors.Wrapf(err, "failed to compile rule for key '%v'", key)
    }

    patterns[k] = m
  }
  return patterns, nil
}

func flattenConfigs(from map[string]interface{}) map[string]interface{} {
  tmp := map[string]interface{}{}
  for k, v := range from {
    recFlattenConfigs(tmp, k, v)    
  }
  return tmp
}

func recFlattenConfigs(to map[string]interface{}, root string, value interface{}) {
  switch v := value.(type) {
  case map[string]interface{}:
    for key, value := range v {
      recFlattenConfigs(to, fmt.Sprintf("%s.%s", root, key), value)
    }
  case []interface{}:
    for i, value := range v {
      recFlattenConfigs(to, fmt.Sprintf("%s.%d", root, i), value)
    }
  default:
    to[root] = value
  }
}
```

Reply:
Thanks for your comments, several things:

- I took the `Unpack` implementation from https://github.com/elastic/beats/blob/master/libbeat/conditions/fields.go#L29-L58, this use case is pretty similar to that
- I need an structure that can be converted back to YAML easily, as the enroll command will end up writing the blacklist, that's why the current one seems convenient.
- It actually works, I'm wondering how since I didn't add a pointer receiver to Unpack, still the map structure is a reference anyway.

Reply:
> I took the Unpack implementation from /libbeat/conditions/fields.go@master#L29-L58, this use case is pretty similar to that

Not exactly. There are differences.

First of all the unpacker interfaces supported by go-ucfg are: BoolUnpacker, ConfigUnpacker, FloatUnpacker, IntUnpacker, StringUnpacker, UintUnpacker, and Unpacker. We indeed miss an Unpacker interface{} accepting `map[string]interface{}`. So using `interface{}` or `*common.Config` as argument to the method is correct.

When using Unpack, you want to persist changes into the method receiver. Therefore you must use a pointer on the method receiver. I doubt your method is even executed by go-ucfg, but then reflection can be a tricky and surprising at times.

If your method is indeed called (which would be a bug in go-ucfg), it could still work if/because:
- pattern variable is initialized with a `map[string]interface{}` (it must not be `nil`)
- maps are always passed by reference. That is, even if `ConfigBlacklistSettings` is used as value receiver, the underlying map will be copied
- if ConfigBlacklistSettings contains default settings, then these will still be there (you should also clean the map).

Some differences between this implementation and the fields one:
- Fields is a struct, no re-type of map[string]interface{}. This shouldn't be a problem though.
- The signature in the Fields example is actually `func (f *Fields) Unpack(to interface{}) error`. It should say `from` not `to` for the parameter (which is indeed confusing).
- the fields example initializes a local empty map overwriting/removing potential defaults: `f.fields = map[string]interface{}{}`. Your sample code does not remove the initial default settings. In the blacklist case the user settings would be merged with the default settings.
- the expand function in fields.go copies the final result into f.fields (`f.fields[key] = value`). The blacklist code copies the final result back into `m` (`m[key] = fmt.Sprintf("%s", value)`), but does not write any settings into `f`.

Are you sure it actually works as expected? I guess there are 2 potential scenarios why you think it works:
- The tests might be running with the defaults + tests themselves would get you same result no matter if defaults or test settings are applied
- go-ucfg doesn't execute the Unpack method, because `*ConfigBlacklistSettings` does not implement the Unpacker interface -> go-ucfg will detect ConfigBlacklistSettings being a `map[string]interface{}` and just merge all config file settings into ConfigBlacklistSettings as is (no flattening of names).

Tip: Add a fmt.Println(...) to Unpack and check if it's actually executed.

I also noticed your tests are somewhat incomplete. The unit tests call NewConfigBlacklist, but there are no unit-test involving the unpacking of `ConfigBlacklistSettings`. Either create separate unit tests for NewConfigBlacklistSettings or change your current unit tests to use config objects. E.g. change `pattern map[string]string` to `config map[string]interface{}`, and have the test create a config object and unpack the config object into `ConfigBlacklistSettings`.

Is it on purpose that user provided configs are merged with the defaults? Or do we rather want to remove defaults if users provide their own settings?


> - I need an structure that can be converted back to YAML easily, as the enroll command will end up writing the blacklist, that's why the current one seems convenient.

Which is ok I think. The problem is not exectly with the type of ConfigBlacklistSettings (well, a shorter name would be nice :) ), but the implementation of Unpack.

Reply:
@urso we also have `FlattenedKeys()` in ucfg, It just return the keys not the value, but in this context it could be already used.

Reply:
Right. We also have a `(MapStr).Flatten` which is similar to `flattenConfigs`.

Reply:
Thanks for your input folks, really appreciated, will do the changes

Reply:
pushed 0f802c3

Reply:
Implementation looks good.